### PR TITLE
Fix `partial_trace` when not removing subsystems

### LINKED ISSUE #7612: Partial trace operation throws up error once it is broken
### Informations

- **Qiskit Aer version**: 0.10.2
- **Python version**: 3.6.9
- **Operating system**: Mac OS 

### What is the current behavior?

See minimal code example below 

### Steps to reproduce the problem
```

# Import Qiskit
from qiskit import QuantumCircuit
import qiskit.quantum_info as qi
from qiskit import Aer, execute

simulator = Aer.get_backend('statevector_simulator', seed = 100)


qc_AB = QuantumCircuit(2)
qc_AB.h(0)
qc_AB.cx(0,1)

sv = execute(qc_AB, simulator).result().get_statevector()

rho_B = qi.partial_trace(sv,[0]) #trace out qubit 0
```

Now If I provide an empty list to trace out then it throws up an error which is expected 
```
rho_B = qi.partial_trace(sv, [])  

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-26-a145dac87eb0> in <module>
----> 1 rho_B = qi.partial_trace(sv,[])

~/PQKs/pqks/lib/python3.6/site-packages/qiskit/quantum_info/states/utils.py in partial_trace(state, qargs)
     56     if isinstance(state, Statevector):
     57         trace_systems = len(state._op_shape.dims_l()) - 1 - np.array(qargs)
---> 58         arr = state._data.reshape(state._op_shape.tensor_shape)
     59         rho = np.tensordot(arr, arr.conj(), axes=(trace_systems, trace_systems))
     60         rho = np.reshape(rho, traced_shape.shape)

ValueError: cannot reshape array of size 4 into shape (2,2,2,2)

```

After the error, if I now recall a previous instruction that successfully ran then the error reappears.

```
rho_B = qi.partial_trace(sv,[0]) #trace out qubit 0

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-27-8cbae0b90ecc> in <module>
----> 1 rho_B = qi.partial_trace(sv,[0])

~/PQKs/pqks/lib/python3.6/site-packages/qiskit/quantum_info/states/utils.py in partial_trace(state, qargs)
     56     if isinstance(state, Statevector):
     57         trace_systems = len(state._op_shape.dims_l()) - 1 - np.array(qargs)
---> 58         arr = state._data.reshape(state._op_shape.tensor_shape)
     59         rho = np.tensordot(arr, arr.conj(), axes=(trace_systems, trace_systems))
     60         rho = np.reshape(rho, traced_shape.shape)

ValueError: cannot reshape array of size 4 into shape (2,2,2,2)

```

It seems as if the `qi.partial_trace` operation is in some state that once it is broken, it cant function again?

If I use `.get_statevector().data` the above issue disappears. 

### What is the expected behavior?

Operation should not be state-bound and work regardless?

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

There were places, such as in `quantum_info.partial_trace` where an
`OpShape` output of `OpShape.remove` was mutated.  In cases where
nothing was removed from the `OpShape`, it returned itself, and
consequently caused lingering issues with the object that owned the
shape.  This ensures a copy is always produced, and adds in a special
case to handle `DensityMatrix` with nothing traced out.

### Details and comments

Fix #7612

### CODE REVIEW

**Comment by jakelishman** [comment]
@ikkoham: would you mind having a look at this?

File: qiskit/quantum_info/states/utils.py [hunk 1]
```diff
@@ -60,6 +60,9 @@ def partial_trace(state, qargs):
         return DensityMatrix(rho, dims=traced_shape._dims_l)
 
     # Density matrix case
+    # Empty partial trace case.
+    if traced_shape == state._op_shape:
```
Comment (ikkoham) [inline qiskit/quantum_info/states/utils.py:64]: Question: Can this condition be replaced by checking an empty list? I think it is straight forward from above your comment. Sorry if there is some deeper reason.

**Review by ikkoham — COMMENTED** [review]
Thank you for fixing the bug. LGTM but I can't understand the reason of one line.

File: qiskit/quantum_info/states/utils.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/quantum_info/states/utils.py]: Sure, it's marginally more efficient to do it your way as well, especially if the shapes are huge.  Done in 1e9e40b.

**Review by ikkoham — APPROVED** [review]
Thank you for changing the code. LGTM

**Pull request merged**